python/hyperonpy.cpp

```diff
@@ -110,6 +117,46 @@ exec_error_t *py_execute(const struct gnd_t* _cgnd, struct vec_atom_t* _args, st
     }
 }
 
+void py_match_(const struct gnd_t *_gnd, const struct atom_t *_atom, lambda_t_binding_array_t callback, void *context) {
+    std::cout << "py_match_ _gnd: "<< _gnd << ", _atom: " << _atom << std::endl;
+    py::object hyperon = py::module_::import("hyperon");
+    py::function call_match_on_grounded_atom = hyperon.attr("call_match_on_grounded_atom");
+    py::object pyobj = static_cast<GroundedObject const *>(_gnd)->pyobj;
+    CAtom catom = (atom_t *)_atom;
+    py::list py_list = call_match_on_grounded_atom(pyobj, catom);
+    size_t size_py_list = py::len(py_list);
+
+    py::dict py_dict = py_list[0];
+    binding_array_t data;
+    binding_t c_binding_t[1];
+    std::string string_array[1];
+    for (auto item : py_dict) {
+        string_array[0] = std::string(py::str(item.first));
+        c_binding_t[0].var = string_array[0].c_str();
+        std::cout << "py_match_ c_binding_t[i].var: " <<  c_binding_t[0].var << std::endl;
+        c_binding_t[0].atom = item.second.attr("catom").cast<CAtom>().ptr;
```

Comment:
The memory issue is here. Python object contains the pointer to the Rust atom. But this instance of the Rust atom is owned by Python object. The pointer is copied inside `binding_t` structure then inside `match_callback` function Rust program behaves as an owner (here https://github.com/h555yand/hyperon-experimental/blob/b71ea42abf2fbd415f2ecf179d51e759c3293236/c/src/atom.rs#L276). I.e. Rust program keeps pointer to the atom and behaves like an owner. At the same time Python atom keeps pointer to the atom and behaves like owner. When space- is finishing processing a query Python atom is cleaned up and frees Rust atom memory. After that Python test tries to print the results and gets access to the memory through the Rust atom which leads to segmentation fault.

Actually there are two ways to fix it. We could copy atom inside `match_callback`. But I consciously used in C API the approach  when caller passes ownership of the atom into the function. For C API it looks logical because on the side of the caller you can choose whether you need additional copy or not. It is also a bit more compact because you don't need freeing created atom when you passed it into the callback. We can remove this additional copying on the Python side, but we need to modify code of Atoms to explicitly encode two states: Rust object is owned and Rust object is not owned. I don't propose doing this now.

Instead I would propose taking the second approach which is used inside `hyperonpy.cpp`: cloning atom before passing it to the C API function. This code fixes the issue:
```suggestion
        c_binding_t[0].atom = atom_clone(item.second.attr("catom").cast<CAtom>().ptr);
```

Reply:
Hi All, thanks for the advice, It fixed segfault.

I've implemented Python unit test.
It works fine for Symbol, Expr, Grounded types but does not work as expected for Variable atom.
For Variable type custom match_ function is not called and I don't is it a bug or a feature.
Could you please advise?

Reply:
Ok, it is normal for a `VariableAtom` because when it is matched with anything else it just gets the other atom as a value. You can check the equality to the added atom instead.

I would not add a code to a `MatchableAtom` itself for the purpose of unit test. I would write a separate class instead and wrap it into the test atom (like it is done in `test_grounded_no_copy` for example.

And last thing: before merging this PR please clean up the code:
- remove prints and replace by logging when it is appropriate
- remove commented code
- remove all shortcuts like this one https://github.com/h555yand/hyperon-experimental/blob/00446b1eb5c1e7d83da87a60dbfbdd73f015af16/python/tests/test_atom.py#L4

Reply:
Hi All,
thanks for the comments.
I've made corresponding changes. Hope it works now.
Regards, Aleksei.